src/solver/td_simplified.ml

```diff
@@ -0,0 +1,233 @@
+
+(** Top down solver that uses the box-operator for widening/narrowing at widening points.
+ * Options:
+ * - solvers.td3.remove-wpoint (default: true): Remove widening points when a variable is and stays stable in iterate. Increases precision of nested loops.*)
+
+open Batteries
+open ConstrSys
+open Messages
+
+module M = Messages
+
+module Base : GenericEqSolver =
+  functor (S:EqConstrSys) ->
+  functor (HM:Hashtbl.S with type key = S.v) ->
+  struct
+    open SolverBox.Warrow (S.Dom)
+    include Generic.SolverStats (S) (HM)
+    module VS = Set.Make (S.Var)
+
+    type solver_data = {
+      infl: VS.t HM.t;
+      rho: S.Dom.t HM.t;
+      wpoint: unit HM.t;
+      stable: unit HM.t;
+    }
+
+    let create_empty_data () = {
+      infl = HM.create 10;
+      rho = HM.create 10;
+      wpoint = HM.create 10;
+      stable = HM.create 10;
+    }
+
+    let print_data data =
+      Logs.debug "|rho|=%d" (HM.length data.rho);
+      Logs.debug "|stable|=%d" (HM.length data.stable);
+      Logs.debug "|infl|=%d" (HM.length data.infl);
+      Logs.debug "|wpoint|=%d" (HM.length data.wpoint)
+
+    let print_data_verbose data str =
+      if Logs.Level.should_log Debug then (
+        Logs.debug "%s:" str;
+        print_data data
+      )
+
+    let solve st vs =
+      let data = create_empty_data () in
+
+      let called = HM.create 10 in
+      let infl = data.infl in
+      let rho = data.rho in
+      let wpoint = data.wpoint in
+      let stable = data.stable in
+
+      let remove_wpoint = GobConfig.get_bool "solvers.td3.remove-wpoint" in
+
+      let () = print_solver_stats := fun () ->
+          print_data data;
+          Logs.info "|called|=%d" (HM.length called);
+          print_context_stats rho
+      in
+
+      let add_infl y x =
+        if tracing then trace "infl" "add_infl %a %a" S.Var.pretty_trace y S.Var.pretty_trace x;
+        HM.replace infl y (VS.add x (HM.find_default infl y VS.empty));
+      in
+
+      let init x =
+        if not (HM.mem rho x) then (
+          new_var_event x;
+          if tracing then trace "init" "init %a" S.Var.pretty_trace x;
+          HM.replace rho x (S.Dom.bot ())
+        )
+      in
+
+      let eq x get set =
+        if tracing then trace "eq" "eq %a" S.Var.pretty_trace x;
+        match S.system x with
+        | None -> S.Dom.bot ()
+        | Some f -> f get set
+      in
+
+      let rec destabilize x =
+        if tracing then trace "destab" "destabilize %a" S.Var.pretty_trace x;
+        let w = HM.find_default infl x VS.empty in
+        HM.replace infl x VS.empty;
+        VS.iter (fun y ->
+            if tracing then trace "destab" "stable remove %a" S.Var.pretty_trace y;
+            HM.remove stable y;
+            destabilize y
+          ) w
+      in
+
+      let rec iterate ?reuse_eq x = (* ~(inner) solve in td3*)
+        let query x y = (* ~eval in td3 *)
```

Comment:
Is there a reason that `query` and `side` are _inside_ `iterate`/`solve` here? This makes `iterate` (which begins a long way down) quite large, which is also apparent from the need to explicitly comment where `iterate` begins. And the `x` argument of `iterate` is always the `x` argument of `query`/`side`, so there's unnecessary duplication.